### DISCUSSION TITLE
Rethinking the Interface

### DISCUSSION THREAD

**Discussion by sor4chi** [sor4chi/hono-storage#70] [General]
The current Hono Storage is not extensible, and I feel that it is impossible to do a Production Release as it is because of its unreasonable abstraction.

This project is a WIP one that hasn't even had a MINOR release yet, but I'm sorry that people are still using it and I'd like to do something about it.

If you have any ideas, please submit them.

**Comment by sor4chi** [comment]
Looking back, I think I was too aware of the expressjs's Multer and made the implementation suffer. I regret that I could have made it simpler.

**Reply by sittiponghaus** [reply]
My feedback might invalid or not related to your ideal roadmap since I'm quite new in this area (was project manager / product owner before).
And I'm not familiar with multer since at that time what I know about coding is front end.

But I'm thinking like multer might not related to today context anymore, what I'm like about it is just the simple validation but when I write API with hono, almost all my validation logic is in zod schema, so for me this might be interesting if I can just simplify the file validation logic inside single middleware.

My other project that I'm working on front end, the backend guy using [tusd](https://github.com/tus/tusd) together with [uppy](https://uppy.io/) for front end, so maybe [tus-node-server](https://github.com/tus/tus-node-server) might help inspire, resumable upload is interesting use case too.

**Reply by sor4chi** [reply]
Thank you.

It may indeed be useful to have [simple validation](https://hono.dev/docs/guides/validation#manual-validator), but we also think that the Hono side already has a formdata validation mechanism and that the hono-storage side should follow that philosophy.

**Comment by sor4chi** [comment]
Current Issues:

- Incomplete Uploads on Validation Failure:
When multiple storage middlewares are applied, if a validation error occurs within the middleware chain, some files may remain partially uploaded. It’s challenging to evaluate whether a middleware is at the end of the chain, making this architecture inherently limited.
- Unclear Key Retrieval Post-Upload:
There is no standardized method for retrieving keys after upload. The abstraction is insufficient, requiring different approaches for different providers.
  - https://github.com/sor4chi/hono-storage/discussions/67
- Lack of Extensibility in Validation:
The current validation logic lacks extensibility. The type definitions have become overly complex, and refactoring is needed to improve maintainability.
- Extending the File Object:
Currently, the File object is being extended, but we aim to minimize this and implement solutions without relying heavily on extending File.
  - https://github.com/sor4chi/hono-storage/issues/69
- Complex Handling of Form Data Fields Containing Text:
The handling of form data fields when they contain text has made the process unnecessarily complicated. Simplifying this logic is a priority.
